clr/commands.py

```diff
@@ -421,28 +421,55 @@ class NamespaceCache:
 
     This allows subsequent calls to `clr help` or `clr completion_*` to be fast.
     Necessary to work around the fact that many clr command namespace modules
-    import the world and initialize state on import.
+    import the world and initialize state on import. The cache is used only for
+    help and completion calls, when actually calling a clr command its spec is
+    always loaded from the module.
     """
 
     def __init__(self):
+        # Place the cache file in the temp dir with a static file name. Data stored in
+        # this file is easily regeneratable in the event of a system restart, but should
+        # be left behind after the process is complete for subsequent clr calls to find.
         tmpdir = os.environ.get("TMPDIR", "/tmp")
         self.cache_fn = os.path.join(tmpdir, "clr_command_cache")
-        # Clr processes are short lived. We don't close the shelve, but are
-        # careful to sync it after writes.
-        self.cache = shelve.open(self.cache_fn)
+
+        try:
+            # Cache is stored on disk as a shelve, but loaded into memory and then
+            # closed right away so that multiple processes won't conflict trying to
```

Comment:
Opening & closing really fast doesn't eliminate the possibility of conflicts in trying to access the cache, just minimizes likelihood of conflicts, right?

(That is, there's no locking mechanism of any kind, we just try to read/write quickly)

Reply:
Very true. I think the collision is pretty low now, but the code works 100% (just a bit slower) if the caching fails every time.